File: C++/longest-alternating-subarray-after-removing-at-most-one-element.cpp
// Time:  O(n)
// Space: O(1)

// dp
class Solution {
public:
    int longestAlternating(vector<int>& nums) {
        int result = 1, up1 = 1, up0 = 1, down1 = 1, down0 = 1;
        for (int i = 0; i + 1 < size(nums); ++i) {
            if (nums[i] < nums[i + 1]) {
                tie(up1, up0, down1, down0) = tuple(down1 + 1, down0 + 1, down0, 1);
            } else if (nums[i] > nums[i + 1]) {
                tie(up1, up0, down1, down0) = tuple(up0, 1, up1 + 1, up0 + 1);
            } else {
                tie(up1, up0, down1, down0) = tuple(up0, 1, down0, 1);
            }
            result = max({result, up1, down1});
        }
        return result;
    }
};

// Time:  O(n)
// Space: O(n)
// prefix sum
class Solution2 {
public:
    int longestAlternating(vector<int>& nums) {
        const auto& cmp = [](int a, int b) {
            return (a > b) - (a < b);
        };

        vector<int> left(size(nums), 1);
        for (int i = 1; i < size(nums); ++i) {
            const auto& diff = cmp(nums[i - 1], nums[i]);
            if (!diff) {
                continue;
            }
            left[i] = (i - 2 >= 0 && cmp(nums[i - 2], nums[i - 1]) == -diff) ? left[i - 1] + 1 : 2;
        }
        vector<int> right(size(nums), 1);
        for (int i = size(nums) - 2; i >= 0; --i) {
            const auto& diff = cmp(nums[i], nums[i + 1]);
            if (!diff) {
                continue;
            }
            right[i] = (i + 2 < size(nums) && cmp(nums[i + 1], nums[i + 2]) == -diff) ? right[i + 1] + 1 : 2;
        }
        int result = ranges::max(left);
        for (int i = 1; i + 1 < size(nums); ++i) {
            const auto& diff = cmp(nums[i - 1], nums[i + 1]);
            if (!diff) {
                continue;
            }
            const auto& l = (i - 2 >= 0 && cmp(nums[i - 2], nums[i - 1]) == -diff) ? left[i - 1] : 1;
            const auto& r = (i + 2 < size(nums) && cmp(nums[i + 1], nums[i + 2]) == -diff) ? right[i + 1] : 1;
            result = max(result, l + r);
        }
        return result;
    }
};

File: Python/longest-alternating-subarray-after-removing-at-most-one-element.py
# Time:  O(n)
# Space: O(1)

# dp
class Solution(object):
    def longestAlternating(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        result = up1 = up0 = down1 = down0 = 1
        for i in range(len(nums)-1):
            if nums[i] < nums[i+1]:
                up1, up0, down1, down0 = down1+1, down0+1, down0, 1
            elif nums[i] > nums[i+1]:
                up1, up0, down1, down0 = up0, 1, up1+1, up0+1
            else:
                up1, up0, down1, down0 = up0, 1, down0, 1
            result = max(result, up1, down1)
        return result


# Time:  O(n)
# Space: O(n)
# prefix sum
class Solution2(object):
    def longestAlternating(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        left = [1]*len(nums)
        for i in xrange(1, len(nums)):
            diff = cmp(nums[i-1], nums[i])
            if not diff:
                continue
            left[i] = left[i-1]+1 if i-2 >= 0 and cmp(nums[i-2], nums[i-1]) == -diff else 2
        right = [1]*len(nums)
        for i in reversed(xrange(len(nums)-1)):
            diff = cmp(nums[i], nums[i+1])
            if not diff:
                continue
            right[i] = right[i+1]+1 if i+2 < len(nums) and cmp(nums[i+1], nums[i+2]) == -diff else 2
        result = max(left)
        for i in xrange(1, len(nums)-1):
            diff = cmp(nums[i-1], nums[i+1])
            if not diff:
                continue
            l = (left[i-1] if i-2 >= 0 and cmp(nums[i-2], nums[i-1]) == -diff else 1)
            r = (right[i+1] if i+2 < len(nums) and cmp(nums[i+1], nums[i+2]) == -diff else 1)
            result = max(result, l+r)
        return result
